zcash_client_sqlite/src/wallet.rs

```diff
@@ -1964,6 +1974,70 @@ pub(crate) fn store_transaction_to_be_sent<P: consensus::Parameters>(
         }
     }
 
+    // Add the transaction to the set to be queried for transaction status. This is only necessary
+    // at present for fully transparent transactions, because any transaction with a shielded
+    // component will be detected via ordinary chain scanning and/or nullifier checking.
+    if !detectable_via_scanning {
+        queue_tx_retrieval(
+            wdb.conn.0,
+            std::iter::once(sent_tx.tx().txid()),
+            TxQueryType::Status,
+            None,
+        )?;
+    }
+
+    Ok(())
+}
+
+pub(crate) fn set_transaction_status(
+    conn: &rusqlite::Transaction,
+    txid: TxId,
+    status: TransactionStatus,
+) -> Result<(), SqliteClientError> {
+    match status {
+        TransactionStatus::TxidNotRecognized | TransactionStatus::NotInMainChain => {
+            // If the transaction is now expired, remove it from the retrieval queue.
+            if let Some(chain_tip) = chain_tip_height(conn)? {
+                conn.execute(
+                    "DELETE FROM tx_retrieval_queue WHERE txid IN (
+                        SELECT txid FROM transactions
+                        WHERE txid = :txid AND expiry_height < :chain_tip_minus_lookahead
+                    )",
+                    named_params![
+                        ":txid": txid.as_ref(),
+                        ":chain_tip_minus_lookahead": u32::from(chain_tip).saturating_sub(VERIFY_LOOKAHEAD)
+                    ],
+                )?;
+            }
+        }
+        TransactionStatus::Mined(height) => {
+            // The transaction has been mined, so we can set its mined height, associate it with
+            // the appropriate block, and remove it from the retrieval queue.
+            let sql_args = named_params![
+                ":txid": txid.as_ref(),
+                ":height": u32::from(height)
+            ];
+
+            conn.execute(
+                "UPDATE transactions
+                 SET mined_height = :height
+                 WHERE txid = :txid",
+                sql_args,
+            )?;
+
+            conn.execute(
+                "UPDATE transactions
+                 SET block = blocks.height
+                 FROM blocks
+                 WHERE txid = :txid
+                 AND blocks.height = :height",
+                sql_args,
+            )?;
+
+            notify_tx_retrieved(conn, txid)?;
```

Comment:
There's a potential hazard here: `notify_tx_retrieved` will remove the txid from `tx_retrieval_queue` regardless of whether the request was for Status or Enhancement. But we haven't updated the `raw` field, so if it was Enhancement, then we'll forget that we need to enhance.

This could happen either with incorrect API usage by the wallet, or as a result of a race condition where the wallet is processing a Status request that gets upgraded to an Enhancement request, and the wallet doesn't notice the upgrade so still calls `set_transaction_status`. (This race condition can happen regardless of `set_transaction_status` and `store_decrypted_tx` being individually transactional.)

Reply:
Even though an `Enhancement` request subsumed a `Status` request, it should never be the case that such an upgrade occurs, because a request for `Enhancement` implies that we don't already have the raw transaction data.

Reply:
I don't follow the argument. Why is it the case that we can't make a Status request and then later, before that is satisfied, see that we need an Enhancement request?

Reply:
We should only ever be making `Status` requests for transactions where we already have complete transaction data (i.e. an `Enhancement` request has already been satisfied.)

Reply:
The comment added in d96bc110555cee26d405896e8e027e055061e654 explains why this is ok.